xarray/core/variable.py

```diff
@@ -376,29 +384,121 @@ def _item_key_to_tuple(self, key):
         else:
             return key
 
+    def _broadcast_indexes(self, key):
+        """
+        Parameters
+        -----------
+        key: One of
+            array
+            a mapping of dimension names to index.
+
+        Returns
+        -------
+        dims: Tuple of strings.
+            Dimension of the resultant variable.
+        indexers: list of integer, array-like, or slice. This is aligned
+            along self.dims.
+        """
+        key = self._item_key_to_tuple(key)  # key is a tuple
+        # key is a tuple of full size
+        key = indexing.expanded_indexer(key, self.ndim)
+        basic_indexing_types = integer_types + (slice,)
+        if all([isinstance(k, basic_indexing_types) for k in key]):
+            return self._broadcast_indexes_basic(key)
+
+        # Detect it can be mapped as an outer indexer
+        # If all key is unlabelled, or
+        # key can be mapped as an OuterIndexer.
+        if all(not isinstance(k, Variable) for k in key):
+            return self._broadcast_indexes_outer(key)
+
+        # If all key is 1-dimensional and there are no duplicate labels,
+        # key can be mapped as an OuterIndexer.
+        dims = []
+        for k, d in zip(key, self.dims):
+            if isinstance(k, Variable):
+                if len(k.dims) > 1:
+                    return self._broadcast_indexes_advanced(key)
+                dims.append(k.dims[0])
+            if not isinstance(k, integer_types):
+                dims.append(d)
+
+        if len(set(dims)) == len(dims):
+            return self._broadcast_indexes_outer(key)
+
+        return self._broadcast_indexes_advanced(key)
+
+    def _broadcast_indexes_basic(self, key):
+        dims = tuple(dim for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        return dims, BasicIndexer(key)
+
+    def _broadcast_indexes_outer(self, key):
+        dims = tuple(k.dims[0] if isinstance(k, Variable) else dim
+                     for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        indexer = []
+        for k in key:
+            if isinstance(k, Variable):
+                k = k.data
+
+            if isinstance(k, integer_types + (slice,)):
+                indexer.append(k)
+            else:
+                k = np.asarray(k)
+                if k.ndim > 1:
+                    raise IndexError("Unlabelled multi-dimensional array "
+                                     "cannot be used for indexing.")
+                indexer.append(k if k.dtype.kind != 'b' else k.nonzero()[0])
```

Comment:
Likewise, prefer unpacking here (I'd probably put this on a separate line like above) -- or use `np.flatnonzero()`.